packages/core/src/application.ts

```diff
@@ -232,6 +308,10 @@ export interface ApplicationConfig {
   servers?: {
     [name: string]: Constructor<Server>;
   };
+  /**
+   * Boot Configuration
+   */
+  bootOptions?: BootOptions;
```

Comment:
> I've moved `bootOptions` to `ApplicationConfig` making it a central place for passing in options. 

I am afraid we are in the process of abandoning this approach, see the already approved pull request https://github.com/strongloop/loopback-next/pull/975 and the original issue https://github.com/strongloop/loopback-next/issues/742 for background information.


Exposing `bootOptions` in `ApplicationConfig` is basically the same as exposing them in `app.boot()` - we are leaking application's internals to application users (e.g. API Connect). Consumers of `Application` class should not (and in most case cannot) know what components or boot options the application requires to work correctly.

Reply:
I'm aware of #742 and my understanding of the issue is to move out the building blocks (Controllers, Components, etc. Arrays that get bound to the application) but keep configuration (ports, host, etc.) in ApplicationConfig. 

I see `bootOptions` as part of Configuration ... It's not something that changes again and again and isn't "building blocks" but config which I think should be a part of `ApplicationConfig`. Not sure how it's different.

Reply:
I strongly disagree that `bootOptions` is a configuration that can be controller by app consumers. It's the app that understands what's its project root and what conventions it uses for laying out controller/provider/repository files in the filesystem, not the consumers of the app!

Quoting from #742 
> - The functionality/features the application should implement - what controllers it contains, what repositories are registered, etc. ("the assembly instructions")
> - The configuration - what port to listen on, what database servers to connect to, etc.

`bootOptions` directly impacts the first point - what functionality/features the application provides. When you change the directory where controllers are located to an invalid value, then suddenly your application looses all public REST APIs and becomes an empty REST server.

Reply:
A user needs to provide us with `bootOptions` at one point or another ... and the best place to identify the `projectRoot` is MyApp ... right? From here ... it is still the user who knows what conventions they follow and this file will usually sit at the top-most level of a project from where you can see the layout. I see this as config for the project. 

That said I do see what you mean by how it impacts the first point ... but won't that always be the case with someone using `boot`? ... if you change your convention from the default and don't tell us, the app will break. 

What approach do you suggest we take if we're not passing in `BootOptions` in `ApplicationConfig` or via `boot()`? I know you proposed using DI for `BootOptions` but I have some concerns about that as follows:
- What happens if someone rebinds `BootOptions` (in a scenario where we support multiple boot calls) ... considering that `Bootstrapper` is a SINGLETON, it'll only keep a copy of the first time `BootOptions` is resolved. That is why even now it doesn't just retrieve the options by injection `ApplicationConfig`.
- Where does this binding take place? MyApp? `index.ts`?

---

- _This point is just a personal opinion and not an issue_. I do think it'll be poor UX to require users to bind Configuration options to MyApp. What happens when there's 10 different components I want to configure? Not to mention new users looking to get started now absolutely must learn DI instead of just being able to write controllers. I see `ApplicationConfig` as the central place where all config should take place for all components and that way a user always knows where to change the config when needed. They can then even load it from JSON file with different config (depending on component ... for production vs. development).